### ISSUE TITLE
More-useful alternative to Batch Norm

### ISSUE THREAD

**Issue by tom-adsfund** [eclipse/deeplearning4j#5840]
Recent work trying to understand why Batch Norm (BN) works shows (https://arxiv.org/pdf/1805.11604.pdf) that there are other ways to achieve the same benefits using simpler methods.

Importantly in practice, BN can't be used with dropout (unstable). And this prevents getting confidence ranges using dropout on networks using BN.

I emailed one of the lead authors of the paper above to learn how to implement the l_1 norm alternative to BN, and here is the strategy:

```
norm = compute_l1_norm(x_1,...,x_n)
mean = compute_mean(x_1,...,x_n)
xHat_i = (x_i - mean) / norm
```

I presume this would be pretty straight forward to implement, because it's simpler than BN. I'm just not familiar with how you'd actually set this up in DL4J.

Aha! Link: https://skymindai.aha.io/features/DL4J-87

**Comment by tom-adsfund** [comment]
The are only referencing ELU and SELU in the related work section because they can be used to achieve an improvement, but there's no connection to the rest of the paper.

How does a custom layer handle the batch?

**Comment by tom-adsfund** [comment]
Thinking about it, it's kind of obvious how the batch is handled. Forget that.

**Comment by tom-adsfund** [comment]
To be honest, I don't know where to start with applying this. Apparently it is like a Batch Norm layer, but I don't understand how to implement one, and the code from that layer is complex. So I'll just have to leave this in case someone else is familiar. I'm not even sure if what's been provided from the author is sufficient to implement it.

**Comment by tom-adsfund** [comment]
@treo Apparently a major problem with ELU is its computational cost. Also my intuition tells me that it's the normalization using statistics from the whole batch that allows the smoothing of the gradients in Batch Norm and the alternative I've mentioned; ELU doesn't use any such statistics.

**Comment by treo** [comment]
My point about the those activation functions was just that they are mentioned in the paper you linked as comparable alteratives, and that dl4j already has them.

**Comment by tom-adsfund** [comment]
Yeah, I guessed that.

Thinking out loud, I'd really like to see the "l1 norm layer" implemented though. Would allow the benefits of batch norm to be combined with dropout, which would also allow confidence ranges.

**Comment by AlexDBlack** [comment]
> I presume this would be pretty straight forward to implement, because it's simpler than BN. I'm just not familiar with how you'd actually set this up in DL4J.

Assuming we're calculating the norms and means on a per-example basis, it should be straightforward to implement as a layer.
If the normalization statistics are global, that might be a bit more complex.

See also layer normalization - https://github.com/deeplearning4j/deeplearning4j/issues/4829

**Comment by tom-adsfund** [comment]
@AlexDBlack from email with the author: "we are computing the L1 norm of the activations along certain directions, much like BatchNorm does with the mean and variance"

**Comment by tom-adsfund** [comment]
We can put this on the back burner, because after a substantial discussion with someone from Cambridge on using dropout for confidence ranges, I think a better approach for me would be to use Google's Trust Score (https://arxiv.org/pdf/1805.11783.pdf).

**Comment by agibsonccc** [comment]
Closing — this issue is over 7 years old with no recent activity. The codebase has undergone a major rewrite. If this is still relevant, please open a new issue with current reproduction steps.
